Flow Fixes

tubbo opened the pull request:
This includes fixes for the category page timeouts, price range display, and order checkouts. It's a draft because I'm not sure if these tests are gonna pass yet..

Commit:
Fix Timeouts On Category Pages

Remove `sku:` setting from the call to `#to_price` to ensure category
pages will render on time. This doesn't seem to have an effect down the
line on checkouts, and I don't think its presence is necessary.

FLOW-17

Commit:
Fix Checkouts and Price Range Display

Decouples the currency from the experience for orders that aren't in a
known experience.

FLOW-17

Commit:
Fix broken tests

Commit:
Fix error with ranges

eric-pigeon commented on app/assets/javascripts/workarea/storefront/flow_io/flow.js:
```diff
@@ -17,75 +17,66 @@ WORKAREA.registerModule('flow', (function () {
+            localizePrices($scope);
 
-            if (!_.isEmpty($("#country-picker"))) {
+            if (!_.isEmpty($("#country-picker", $scope))) {
```
country picker isn't scoped, its a global html id

tubbo replied:
I don't necessarily want to re-init the country picker if the `$scope` isn't `document`, though...

eric-pigeon replied:
Exactly, so $('#country-picker') won't be empty after the initial module init, but $('#country-picker', NEW_PDP_DOM) will always be empty

tubbo replied:
oh that's true

eric-pigeon commented on app/models/workarea/flow_io/local_item.rb:
```diff
@@ -23,6 +33,19 @@ def to_price(quantity = 1)
+      # TODO determine if we still need the original
+      #
+      #
+      def to_flow_label(quantity = 1, &block)
```
I think this method can be deleted now

Commit:
Write tests for Pricing::Collection changes

Commit:
Fix naming in method

Commit:
Bring back the sku so collection tests will pass, test this on QA for performance

Commit:
Remove to_flow_label since it is no longer used

tubbo commented on app/models/workarea/flow_io/localized_price.rb:
```diff
@@ -12,14 +12,24 @@ class LocalizedPrice
        def to_price
          Workarea::Pricing::Price.new(
-           sku: self.local_item.sku.clone, # clone the sku so this price isn't added to #prices on real record
+           sku: sku,
```
performance seems to be OK on https://plugins-qa.demo.workarea.com with this added...I think it was the `#clone` causing all the trouble.

eric-pigeon replied:
Using sku causes this price to get persisted on the sku, mongoid sucks

tubbo replied:
I don't see the return value of `#to_price` actually being saved anywhere, so I don't think this will have any harmful effects.

tubbo replied:
According to the tests, replacing this with a proxy object that behaves like a `Pricing::Price` is also on the table, the only downside there is that any decorations made to `Pricing::Price` will also need to be made to this `FlowIo::PricingPriceProxy`. This might be a more performant way to go, but has a cognitive cost. I'm personally in favor of leaving it as a `Pricing::Price` since there's nothing that will ever save this object in the Flow code base, but if you think that could cause problems down the road, I'm all ears.

bencrouse replied:
Quick test:

```
irb(main):001:0> sku = Workarea::Pricing::Sku.sample
MONGODB | [13] localhost:27017 #2 | my_app_development.count | STARTED | {"count"=>"workarea_pricing_skus", "query"=>{}, "lsid"=>{"id"=><BSON::Binary:0x70359377325060 type=uuid data=0xef29dcb7317542fb...>}}
MONGODB | [13] localhost:27017 | my_app_development.count | SUCCEEDED | 0.004s
MONGODB | [14] localhost:27017 #2 | my_app_development.find | STARTED | {"find"=>"workarea_pricing_skus", "filter"=>{}, "skip"=>386, "sort"=>{"_id"=>1}, "limit"=>1, "singleBatch"=>true, "lsid"=>{"id"=><BSON::Binary:0x70359377325060 type=uuid data=0xef29dcb7317542fb...>}}
MONGODB | [14] localhost:27017 | my_app_development.find | SUCCEEDED | 0.009s
=> #<Workarea::Pricing::Sku _id: 992321835-X, created_at: 2020-03-03 16:55:39 UTC, updated_at: 2020-03-03 16:55:39 UTC, created_by_site_id: nil, active_segment_ids: {"en"=>[]}, active: {"en"=>true}, on_sale: false, discountable: true, tax_code: "001", msrp: {"cents"=>5826.0, "currency_iso"=>"USD"}>
irb(main):002:0> Workarea::Pricing::Price.new(sku: sku, min_quantity: 2, regular: 5.to_m)
=> #<Workarea::Pricing::Price _id: 5e68e3be021823835db80a41, created_at: nil, updated_at: nil, created_by_site_id: nil, active_segment_ids: {"en"=>[]}, active: {"en"=>true}, min_quantity: 2, regular: {:cents=>500.0, :currency_iso=>"USD"}, sale: nil, on_sale: false>
irb(main):003:0> Workarea::Pricing::Price.new(sku: sku, min_quantity: 2, regular: 5.to_m).persisted?
=> false
```

Another way to prevent any persisting would be to call `freeze` on the instance.

Commit:
Remove test for msrp_for_experience

Commit:
Localize regular price

Commit:
Fix broken Order tests

Commit:
Fix storefront system test

Commit:
Fix storefront system tests by allowing missing data in flow pricing overrides

Commit:
Remove test for functionality that is no longer supported

tubbo commented:
this includes work for both FLOW-17 and FLOW-18, but i needed FLOW-17 in so i could actually test FLOW-18. sorry for the confusion.

Commit:
Fix country picker scoping

bencrouse commented on app/views/workarea/storefront/products/_pricing.html.haml:
```diff
@@ -1,43 +1,55 @@
+          %span{ data: { flow_price_attr: product.original_min_price_attribute, local_prices: product.localized_original_min_prices.to_json } }
             = number_to_currency product.original_min_price
-
-  = append_partials('storefront.product_pricing', product: product)
```
Why remove the append point?

bencrouse commented on test/integration/workarea/storefront/flow_io_webhooks/order_integration_test.rb:
```diff
@@ -101,17 +101,6 @@ def test_order_not_found
           )
         end
 
-        def test_order_placed_in_different_currency
```
Seems like a valid test?

bencrouse commented on test/integration/workarea/storefront/search_integration_test.decorator:
```diff
@@ -1,51 +0,0 @@
-module Workarea
-  decorate Storefront::SearchIntegrationTest, with: :flow_io do
```
Why delete a whole test file?

tubbo replied:
price faceting based on your currency doesn't work anymore, it will only work based on the known experience.

bencrouse commented on test/models/workarea/pricing/collection_test.decorator:
```diff
@@ -2,139 +2,351 @@ module Workarea
+      setup :setup_models
+    end
+
+    def setup_models
```
Why do `setup_models` if the method is empty?

bencrouse reviewed (commented):
Not clear to me why so many tests are being removed. Not many added, but I wanna ship this.

bencrouse reviewed (changes requested):
Looks like a flow test failed in the build...

tubbo commented:
tests are fixed, i'm still working on the plugins-qa issues

bencrouse commented:
@tubbo so are you looking to include the plugins-qa issues in this PR before merging?

tubbo commented:
yes, I’ll be getting that done today.

> On Mar 18, 2020, at 8:51 AM, Ben Crouse <notifications@github.com> wrote:
> 
> 
> @tubbo <https://github.com/tubbo> so are you looking to include the plugins-qa issues in this PR before merging?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/workarea-commerce/workarea-flow-io/pull/29#issuecomment-600604749>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAA3TAS63WQE4DGHWSYGRJ3RIC7V5ANCNFSM4LEROTBA>.
>

Commit:
Fix errors on category pages

tubbo commented:
@eric-pigeon I think this is just in the context of the `Pricing::Collection`. If there _are_ prices being saved, I certainly can't see them...

```ruby
product = Workarea::Catalog::Product.find_by(slug: 'glint-eastwood-blazer') # a product i know i bought
Workarea::Pricing::Sku.find(product.skus).flat_map(&:prices).all? { |price| price.regular.currency == 'USD' } # => true
```

bencrouse commented on app/models/workarea/pricing/collection.decorator:
```diff
@@ -1,81 +1,111 @@
+        .map { |p| [p.sku, p.regular] }
+        .sort_by { |(_sku, regular)| regular }
         .first
+        &.first
```
If you've got an array here to `map` and `sort_by`, it won't be `nil` when you call `first`.

tubbo replied:
not sure I understand...

bencrouse commented:
![gif-elon-musk-dancing-forward-and-backward-on-stage-shanghai](https://user-images.githubusercontent.com/165111/77165069-00382f00-6a88-11ea-9658-3ce3b2ca5c14.gif)

me trying to code review this every day

Commit:
Fix Incorrect Checkout After Placing Order

Always read from the `_f60_session` cookie when setting the Flow session
in the backend. This will be set by FlowJS and thus be more accurate
than our previous method of falling through a header and cookie.

FLOW-17

Commit:
Fix currency display

Commit:
Fix Flow Order Navigation Not Showing

Commit:
Ignore XSS Vulnerability CVE since it won't be patched in Rails 5.1
